src/mlpack/methods/ann/layer/scaled_dot_product_attention_impl.hpp

```diff
@@ -0,0 +1,179 @@
+template <typename eT>
+void ScaledDotProductAttention<InputDataType, OutputDataType>::
+Forward(const arma::Mat<eT>& input, arma::Mat<eT>& output)
+{
+  typedef typename arma::Cube<eT> CubeType;
+  typedef typename arma::Mat<eT> MatType;
+
+  Log::Assert(input.n_rows == embedDim * (tLen + 2 * sLen));
+
+  const size_t bsz = input.n_cols;
+  const size_t qStart = 0, qEnd = embedDim * tLen - 1;
+  const size_t kStart = qEnd + 1, kEnd = kStart + embedDim * sLen - 1;
+  const size_t vStart = kEnd + 1, vEnd = vStart + embedDim * sLen - 1;
+  output.set_size(embedDim * tLen, bsz);
+
+  MatType q = input.submat(qStart, 0, qEnd, bsz - 1);
+  MatType k = input.submat(kStart, 0, kEnd, bsz - 1);
+  MatType v = input.submat(vStart, 0, vEnd, bsz - 1);
+
+  CubeType query(q.memptr(), embedDim, tLen, bsz);
+  CubeType key(k.memptr(), embedDim, sLen, bsz);
+  CubeType value(v.memptr(), embedDim, sLen, bsz);
+
+  query /= std::sqrt(embedDim);
+  CubeType scores = CubeMultiply(key, query, 1, 0);
+
+  if (!attnMask.is_empty())
+  {
+    if (attnMask.n_rows != sLen || attnMask.n_cols != tLen)
+    {
+      Log::Fatal << "The size of the 2D `attn_mask` is not correct."
+                 << std::endl;
+    }
+    scores.each_slice() += attnMask;
+  }
+  if (!keyPaddingMask.is_empty())
+  {
+    if (keyPaddingMask.n_rows != sLen || keyPaddingMask.n_cols != 1)
+    {
+      Log::Fatal << "The size of the `keyPaddingMask` is not correct."
+                 << std::endl;
+    }
+    scores.each_slice() += arma::repmat(keyPaddingMask, 1, tLen);
+  }
+
+  attnOut.set_size(sLen, tLen, bsz);
+  softmaxOutput.set_size(sLen, tLen, bsz);
+  for (size_t i = 0; i < bsz; ++i)
+  {
+    softmax.Forward(scores.slice(i), softmax.OutputParameter());
+    softmaxOutput.slice(i) = softmax.OutputParameter();
+    dropout.Forward(softmax.OutputParameter(), attnOut.slice(i));
+  }
+  scores = CubeMultiply(value, attnOut, 0, 0);
+  for (size_t i = 0; i < bsz; ++i)
+  {
+    output.col(i) = arma::vectorise(scores.slice(i));
+  }
+}
+
+template <typename InputDataType, typename OutputDataType>
+template <typename eT>
+void ScaledDotProductAttention<InputDataType, OutputDataType>::
+Backward(const arma::Mat<eT>& input,
+         const arma::Mat<eT>& gy,
+         arma::Mat<eT>& g)
+{
+  typedef typename arma::Cube<eT> CubeType;
+  typedef typename arma::Mat<eT> MatType;
+
+  const size_t bsz = gy.n_cols;
+  const size_t qStart = 0, qEnd = embedDim * tLen - 1;
+  const size_t kStart = qEnd + 1, kEnd = kStart + embedDim * sLen - 1;
+  const size_t vStart = kEnd + 1, vEnd = vStart + embedDim * sLen - 1;
+  g.set_size(embedDim * (tLen + 2 * sLen), bsz);
+
+  MatType q = input.submat(qStart, 0, qEnd, bsz - 1);
+  MatType k = input.submat(kStart, 0, kEnd, bsz - 1);
+  MatType v = input.submat(vStart, 0, vEnd, bsz - 1);
+
+  CubeType query(q.memptr(), embedDim, tLen, bsz);
+  CubeType key(k.memptr(), embedDim, sLen, bsz);
+  CubeType value(v.memptr(), embedDim, sLen, bsz);
```

Comment:
How do you think, is it possible to get rid of the code duplication here? I mean this part of code is identical to the corresponding code in `Forward()`.

Reply:
I've changed the code a little bit. I am not taking `key` and `value` as input, rather as arguments. I strongly feel this way is better and easy. 
The memory failures are unrelated as you might know.